src/codec.rs

```diff
@@ -488,29 +550,122 @@ impl<T, X> Encode for X where
+		None
+	}
 }
+
 impl<T> WrapperTypeDecode for Box<T> {
 	type Wrapped = T;
+
+	fn decode_wrapped<I: Input>(input: &mut I) -> Option<Result<Self, Error>> where Self::Wrapped: Decode {
+		if let Err(error) = input.descend_ref() {
+			return Some(Err(error));
+		}
+
+		// Placement new is not yet stable, but we can just manually allocate a chunk of memory
+		// and convert it to a `Box` ourselves.
+		//
+		// The explicit types here are written out for clarity.
+		//
+		// TODO: Use `Box::new_uninit` once that's stable.
+		let layout = core::alloc::Layout::new::<MaybeUninit<T>>();
+
+		let ptr: *mut MaybeUninit<T> = if layout.size() == 0 {
+			core::ptr::NonNull::dangling().as_ptr()
+		} else {
+
+			// SAFETY: Layout has a non-zero size so calling this is safe.
+			let ptr: *mut u8 = unsafe {
+				crate::alloc::alloc::alloc(layout)
+			};
+
+			if ptr.is_null() {
+				crate::alloc::alloc::handle_alloc_error(layout);
+			}
+
+			ptr.cast()
+		};
+
+		// SAFETY: Constructing a `Box` from a piece of memory allocated with `std::alloc::alloc`
+		//         is explicitly allowed as long as it was allocated with the global allocator
+		//         and the memory layout matches.
+		//
+		//         Constructing a `Box` from `NonNull::dangling` is also always safe as long
+		//         as the underlying type is zero-sized.
+		let mut boxed: Box<MaybeUninit<T>> = unsafe { Box::from_raw(ptr) };
+
+		if let Err(error) = T::decode_into(DecodeContext::new(), input, &mut boxed) {
+			return Some(Err(error));
+		}
+
+		// Decoding succeeded, so let's get rid of `MaybeUninit`.
+		let ptr: *mut MaybeUninit<T> = Box::into_raw(boxed);
+		let ptr: *mut T = ptr.cast();
+
+		// SAFETY: `MaybeUninit` doesn't affect the memory layout, so casting the pointer back
+		//         into a `Box` is safe.
+		let boxed: Box<T> = unsafe { Box::from_raw(ptr) };
+
+		input.ascend_ref();
+		Some(Ok(boxed))
+	}
 }
+
 impl<T> WrapperTypeDecode for Rc<T> {
 	type Wrapped = T;
+
+	fn decode_wrapped<I: Input>(input: &mut I) -> Option<Result<Self, Error>> where Self::Wrapped: Decode {
+		// TODO: This is inefficient; use `Rc::new_uninit` once that's stable.
+		Some(Box::<T>::decode(input).map(|output| output.into()))
+	}
 }
+
 #[cfg(target_has_atomic = "ptr")]
 impl<T> WrapperTypeDecode for Arc<T> {
 	type Wrapped = T;
+
+	fn decode_wrapped<I: Input>(input: &mut I) -> Option<Result<Self, Error>> where Self::Wrapped: Decode {
+		// TODO: This is inefficient; use `Arc::new_uninit` once that's stable.
+		Some(Box::<T>::decode(input).map(|output| output.into()))
+	}
 }
 
 impl<T, X> Decode for X where
 	T: Decode + Into<X>,
 	X: WrapperTypeDecode<Wrapped=T>,
 {
+	#[inline]
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
-		input.descend_ref()?;
-		let result = Ok(T::decode(input)?.into());
-		input.ascend_ref();
-		result
-	}
+		// First try to see if a specialized implementation exists.
+		if let Some(result) = Self::decode_wrapped(input) {
+			return result;
+		}
```

Comment:
Maybe, we could simply call: `Self::decode_wrapped(input)`, and provide the `default` implementation with the code below?